src/target/adiv5.c

```diff
@@ -768,8 +768,7 @@ adiv5_access_port_s *adiv5_new_ap(adiv5_debug_port_s *dp, uint8_t apsel)
 			ap.base |= (uint64_t)adiv5_ap_read(&ap, ADIV5_AP_BASE_HIGH) << 32U;
 		}
 		/* Check the Debug Base Address register for not-present. See ADIv5 Specification C2.6.1 */
-		if (base_flags == (ADIV5_AP_BASE_FORMAT_ADIV5 | ADIV5_AP_BASE_PRESENT_NO_ENTRY) ||
-			(!(ap.flags & ADIV5_AP_FLAGS_64BIT) && (uint32_t)ap.base == ADIV5_AP_BASE_NOT_PRESENT)) {
+		if (!(ap.flags & ADIV5_AP_FLAGS_64BIT) && (uint32_t)ap.base == ADIV5_AP_BASE_NOT_PRESENT) {
```

Comment:
The specification for ADIv5 makes it very clear the check that's being removed here is required, as there are no validity requirements around `BASE.BASEADDR[31:12]` - any value in those bits is valid - and only bits [11:0] have validity requirements placed on them.

What is the cause of making this change? We don't see a noted reasoning for it in the PR or anywhere obvious in the commit messages. It might be helpful to attach a ROM table dump (`blackmagic -tv 5`) from the part so we can understand what's going on in the BASE field for it to help us square what's going on here against other parts and the specification.

Reply:
I think let's start from the root, ie. the data that is being read from the device. Here's the -tv5 output:

```
Black Magic Debug App v1.10.0-852-gf650fa8f-dirty
 for Black Magic Probe, ST-Link v2 and v3, CMSIS-DAP, J-Link and FTDI (MPSSE)
Using 1d50:6018 81D15AA3 Black Magic Debug
 Black Magic Probe v1.10.2
Running in Test Mode
Target voltage: 3.3V Volt
Speed set to 1.951MHz for SWD
Switching out of dormant state into SWD
Handling SWD multi-drop, TARGETID 0x1bb8202f
DP DPIDR 0x6ba02477 (v2 rev6) designer 0x43b partno 0xba
TARGETID 0x1bb8202f designer 0x17 partno 0xbb82
AP   0: IDR=84770001 CFG=00000000 BASE=f0000000 CSW=a3800040 (AHB3-AP var0 rev8)
Halt via DHCSR(01030003): success after 6ms
ROM: Table BASE=0xf0000000 SYSMEM=1, Manufacturer 43b Partno 4c1 (PIDR = 0x00000004000bb4c1)
ROM: Table BASE=0xe00ff000 SYSMEM=1, Manufacturer 43b Partno 4c0 (PIDR = 0x00000004000bb4c0)
 0 0xe000e000: Generic IP component - Cortex-M0 SCS (System Control Space) (PIDR = 0x00000004000bb008 DEVTYPE = 0x00 ARCHID = 0x0000)
 -> cortexm_probe
CPUID 0x410cc601 (M0+ var 0 rev 1)
Calling msp432p4_probe
Calling mspm0_probe
 1 0xe0001000: Generic IP component - Cortex-M0 DWT (Data Watchpoint and Trace) (PIDR = 0x00000004000bb00a DEVTYPE = 0x00 ARCHID = 0x0000)
 2 0xe0002000: Generic IP component - Cortex-M0 BPU (Breakpoint Unit) (PIDR = 0x00000004000bb00b DEVTYPE = 0x00 ARCHID = 0x0000)
 ROM: Table END
1 Entry 0x1002 -> Not present
2 Entry 0x50402002 -> Not present
3 Entry 0x10000002 -> Not present
ROM: Table END
AP   1: IDR=002e0001 (Unknown var0 rev0)
0 0x00000000: 0x00000000 <- does not match preamble (0xb105000d)
AP   2: IDR=002e0000 (Unknown var0 rev0)
0 0x00000000: 0x00000000 <- does not match preamble (0xb105000d)
AP   3: IDR=002e0003 (Unknown var0 rev0)
0 0x00000000: 0x01010101 <- does not match preamble (0xb105000d)
AP   4: IDR=002e0002 (Unknown var0 rev0)
0 0x00000000: 0x00000000 <- does not match preamble (0xb105000d)
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
SWD access resulted in no response
***  1   MSPM0 M0+
RAM   Start: 0x20000000 length = 0x1000
Flash Start: 0x00000000 length = 0x10000 blocksize 0x400
```

Also the interesting part with -tv15
```
TARGETID 0x1bb8202f designer 0x17 partno 0xbb82
Write CTRL: 0x00000000
Read STATUS: 0x00000040
Write CTRL: 0x50000000
Read STATUS: 0xf0000040
Read AP 0 IDR: 0x84770001
AP   0: IDR=84770001Read AP 0 CFG: 0x00000000
Read AP 0 CSW: 0x03800052
Read AP 0 BASE: 0xf0000002
 CFG=00000000 BASE=f0000000 CSW=a3800040 (AHB3-AP var0 rev8)
Read STATUS: 0xf0000040
Write AP 0 CSW: 0xa3800042
Write AP 0 TAR: 0xe000edf0
Write AP 0 DRW: 0xa05f0001
Read RDBUFF: 0xf0000002
Write CTRL: 0xf0fff040
Write AP 0 DRW: 0xa05f0003
Read AP 0 DRW: 0xf0000002
Write CTRL: 0xf0fff040
Write AP 0 DRW: 0xa05f0003
```

The BASE.P bit is reported as zero. Although the address points to a valid ROM table and obviously skipping this AP makes the device unavailable. My experience with ADI etc. is just what I gathered while coding this. Reading the ADIv5 docs, the meaning of BASE.P field is sowewhat a mystery to me as the description in the docs is quite short. They talk about 'debug components' then say that the flag denotes the presence of 'debug entry' which.. is the same thing or not? So I will not argue that simply ignoring the bit is the correct way as the device might be buggy as well. But one thing is I wonder what other devices (with a ROM table) report. I noticed that this part of code was rewritten somewhat recently (at one point mspm0 support I wrote stopped working after a merge with the main) and wonder how much real-world testing it has gotten. The other thing is how to come around this issue if probing is supposed to ignore devices with zero BASE.P values. Check manufacturer/partnum and work around?

Somewhat unrelated, but doubt it warrants a separate issue. I noticed there's a bug in the original (unmodified) code. The function first discards lower bits of `ap.base` and later tests if address is equal to NOTPRESENT (in legacy format). This is not correct as ap.base is already mangled and will never be equal to 0xffff.ffff.

I think there should be a test right after reading BASE register, like so:
```
		ap.base = adiv5_ap_read(&ap, ADIV5_AP_BASE_LOW);
		if (!(ap.flags & ADIV5_AP_FLAGS_64BIT) && ap.base == ADIV5_AP_BASE_NOT_PRESENT) {
			DEBUG_INFO(" -> Not Present (Legacy)\n");
			return NULL;
		}
```

Reply:
There's a lot here, so we're going to answer this in parts:

> The BASE.P bit is reported as zero. Although the address points to a valid ROM table and obviously skipping this AP makes the device unavailable. My experience with ADI etc. is just what I gathered while coding this. Reading the ADIv5 docs, the meaning of BASE.P field is sowewhat a mystery to me as the description in the docs is quite short. They talk about 'debug components' then say that the flag denotes the presence of 'debug entry' which.. is the same thing or not? So I will not argue that simply ignoring the bit is the correct way as the device might be buggy as well. But one thing is I wonder what other devices (with a ROM table) report. I noticed that this part of code was rewritten somewhat recently (at one point mspm0 support I wrote stopped working after a merge with the main) and wonder how much real-world testing it has gotten. The other thing is how to come around this issue if probing is supposed to ignore devices with zero BASE.P values. Check manufacturer/partnum and work around?

So, that bit in non-legacy mode performs the exact same function as `ADIV5_AP_BASE_NOT_PRESENT`, except with inverted logic - those bottom two bits of BASE being 2 means that the device is telling the debugger that there are no entries (no ROM table, no debug components, nothing) present on the AP. This would ordinarily be correct and is in fact exactly how parts like the STM32H5's and STM32H7's and the LPC4370 all tell us of APs that should be ignored. Failure to ignore those APs can very well crash the parts.

For example:
```
❯ ./blackmagic -s 8BB20695 -tpv 5
Black Magic Debug App v1.10.0-844-ga44f08695-dirty
 for Black Magic Probe, ST-Link v2 and v3, CMSIS-DAP, J-Link and FTDI (MPSSE)
Using 1d50:6018 8BB20695 Black Magic Debug
 Black Magic Probe v1.10.0-843-g070d59f21-dirty
Running in Test Mode
Target voltage: 3.2V Volt
Speed set to 1.951MHz for SWD
Switching out of dormant state into SWD
SWD access resulted in no response
Deprecated JTAG to SWD sequence
Handling SWD multi-drop, TARGETID 0x04830041
DP DPIDR 0x6ba02477 (v2 rev6) designer 0x43b partno 0xba
TARGETID 0x04830041 designer 0x20 partno 0x4830
AP   0: IDR=84770001 CFG=00000000 BASE=e00fe000 CSW=a3800040 (AHB3-AP var0 rev8)
Halt via DHCSR(00030003): success after 28ms
ROM: Table BASE=0xe00fe000 SYSMEM=1, Manufacturer 020 Partno 483 (PIDR = 0x00000000000a0483)
ROM: Table BASE=0xe00ff000 SYSMEM=1, Manufacturer 43b Partno 4c7 (PIDR = 0x00000004000bb4c7)
 0 0xe000e000: Generic IP component - Cortex-M4 SCS (System Control Space) (PIDR = 0x00000004000bb00c DEVTYPE = 0x00 ARCHID = 0x0000)
 -> cortexm_probe
CPUID 0x411fc272 (M7 var 1 rev 2)
Calling stm32f1_probe
Calling stm32f4_probe
Calling stm32h5_probe
Calling stm32h7_probe
 1 0xe0001000: Generic IP component - Cortex-M3 DWT (Data Watchpoint and Trace) (PIDR = 0x00000004000bb002 DEVTYPE = 0x00 ARCHID = 0x0000)
 2 0xe0002000: Generic IP component - Cortex-M7 FBP (Flash Patch and Breakpoint) (PIDR = 0x00000004000bb00e DEVTYPE = 0x00 ARCHID = 0x0000)
 3 0xe0000000: Generic IP component - Cortex-M3 ITM (Instrumentation Trace Module) (PIDR = 0x00000004000bb001 DEVTYPE = 0x00 ARCHID = 0x0000)
 4 Entry 0xfff41002 -> Not present
 5 Entry 0xfff42002 -> Not present
 ROM: Table END
1 0xe0041000: Debug component - Cortex-M7 ETM (Embedded Trace) (PIDR = 0x00000004001bb975 DEVTYPE = 0x13 ARCHID = 0x4a13)
2 0xe0043000: Debug component - CoreSight CTI (Cross Trigger) (PIDR = 0x00000004004bb906 DEVTYPE = 0x14 ARCHID = 0x0000)
3 Entry 0x1ff02002 -> Not present
ROM: Table END
AP   1: IDR=84770001 -> Not Present
AP   2: IDR=54770002 CFG=00000000 BASE=e00e0000 CSW=80000040 (APB2/3-AP var0 rev5)
ROM: Table BASE=0xe00e0000 SYSMEM=0, Manufacturer 020 Partno 483 (PIDR = 0x00000000000a0483)
0 0xe00e1000: 0x00000000000a0000 Non-ARM component ignored
  -> designer: 20, part no: 0
1 Entry 0x2002 -> Not present
2 0xe00e3000: Debug component - CoreSight SWO (Single Wire Output) (PIDR = 0x00000004002bb914 DEVTYPE = 0x11 ARCHID = 0x0000)
3 Entry 0x4002 -> Not present
4 Entry 0x5002 -> Not present
ROM: Table BASE=0xe00f0000 SYSMEM=0, Manufacturer 020 Partno 001 (PIDR = 0x00000000000a0001)
 0 0xe00f1000: Debug component - CoreSight CTI (Cross Trigger) (PIDR = 0x00000004005bb906 DEVTYPE = 0x14 ARCHID = 0x0000)
 1 Entry 0x2002 -> Not present
 2 0xe00f3000: Debug component - CoreSight CSTF (Trace Funnel) (PIDR = 0x00000004003bb908 DEVTYPE = 0x12 ARCHID = 0x0000)
 3 0xe00f4000: Debug component - CoreSight TMC (Trace Memory Controller) (PIDR = 0x00000004001bb961 DEVTYPE = 0x32 ARCHID = 0x0000)
 4 0xe00f5000: Debug component - CoreSight TPIU (Trace Port Interface Unit) (PIDR = 0x00000004005bb912 DEVTYPE = 0x11 ARCHID = 0x0000)
 5 Entry 0x6002 -> Not present
 6 Entry 0x7002 -> Not present
 7 Entry 0x8002 -> Not present
 ROM: Table END
ROM: Table END
```

For which that AP 1 entry:
```
Read AP 1 IDR: 0x84770001
AP   1: IDR=84770001
Read AP 1 CFG: 0x00000000
Read AP 1 CSW: 0x43800042
Read AP 1 BASE: 0x00000002
 -> Not Present
 ```

If BMD tries to read that AP, it crashes the part.

Now, in this case, because the BASEADDR is non-zero (even if a very strange value.. this part is doing a lot of weird things), we can probably get away with an errata check that base matches this part's address value, and skip the skip.

> Somewhat unrelated, but doubt it warrants a separate issue. I noticed there's a bug in the original (unmodified) code. The function first discards lower bits of ap.base and later tests if address is equal to NOTPRESENT (in legacy format). This is not correct as ap.base is already mangled and will never be equal to 0xffff.ffff.

Having just checked that logic, this is what we get for moving stuff around and having next to no parts to test that use the legacy non-presence value. You are correct this is wrong, what probably needs to happen is that
```c
		/* Make sure we only pay attention to the base address, not the presence and format bits */
		ap.base &= ADIV5_AP_BASE_BASEADDR;
```
gets moved down to just after the presence check, which then makes any tests done in the presence check if block a bit simpler.

Note that the comment inside that block documents another part (STM32MP15) which reads back with a non-legacy non-presence marker on one of its PAs, and another where reading that AP anyway precipitates a crash.

Reply:
Please see the latest (29e45dd) update. I think I fixed the legacy no-entry check and made so the discovery passes for TI parts when base-addr is non-zero. Don't really like using manufacturer number in that function, but there are similar ifs in a adiv5_new_ap-calling function, so maybe this is not too ugly. Should I check for partnums as well? But the issue may apply for the whole family (or may not, can't tell for other than MSPM0L1306 variant).

Reply:
The general idea looks sound and as you note, there's precedent for checking the part manufacturer code there so that's perfectly fine. The main thing we're suggesting below is a reduction in the repeated return condition by condensing all the checks. It is a suggestion only as technically the code you've come up with works.